[quant][pt2e] QNNPackQuantizer support annotation for resnet18

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98560
* __->__ #98507

Summary:
This PR adds annotation support for conv2d relu, linear, maxpool2d, add and add relu so
that we can successfully quantize resnet18 with the prepare_pt2e_quantizer API and get the same result
as fx graph mode quantization

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels.test_resnet18_with_quantizer_api

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e] QNNPackQuantizer support annotation for resnet18

Summary:
This PR adds annotation support for conv2d relu, linear, maxpool2d, add and add relu so
that we can successfully quantize resnet18 with the prepare_pt2e_quantizer API and get the same result
as fx graph mode quantization

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels.test_resnet18_with_quantizer_api

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] QNNPackQuantizer support annotation for resnet18"


Summary:
This PR adds annotation support for conv2d relu, linear, maxpool2d, add and add relu so
that we can successfully quantize resnet18 with the prepare_pt2e_quantizer API and get the same result
as fx graph mode quantization

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels.test_resnet18_with_quantizer_api

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -235,32 +256,168 @@ def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
         ops = self.get_supported_operator_for_operator_spec(global_spec)
-        for op in ops:
-            self._annotate_op(model, op, global_spec)
+        # annotate the nodes from last to first since the matching is in the reversed order
```
this one I think is worth fixing now

jerryzh168 replied:
there are two things:
(1) matching each pattern in reverse order:
this is talking about in each _annotate_op function, we start matching from the last node

I think this one is common practice for matching, the reason is that if we match in the normal order, one node can be used by multiple users, and we'd be doing a much more complicated subgraph matching since it can flow from one node to a lot of subgraphs, and we will need to have rules on what to do when we have multiple matching subgraphs etc. Conversely, if we match from output to input, there will only be one path, this will be a easier case to handle. Also this might be replaced by SubgraphMatcher in the future

(2). doing annotation for nodes in reverse order:

also I actually started with matching execution order, and we will need more rules to make the match correct, 
current annotation rule: **every node can only be annotated once.**

e.g. 
x -> conv -> relu -> output
step 1. we handling conv node, we will try to match "conv" and "conv - relu" pattern, then we matched "conv" node, and do the annotation, we also add a "_annotated": True to the target_dtype_info dict, which means it will not be annotated again
step 2. when we are handling relu node, we will be able to match "conv - relu" pattern, but since conv is already annotated, we will not update the annotation unless we change the rule

this will give us incorrect result, since we probably want to annotate "conv - relu" instead of just "conv"
if we want to get the correct behavior, we need to update the rule to: **we can annotate one node multiple times, and the annotation survived in the end is whichever comes last** I think this rule will make the annotation behavior less deterministic and harder to understand, wdyt?

vkuzo replied:
Right, for how an FX graph is defined, matching linear patterns in natural order will have a cleaner config and slightly more complex code, and matching in reverse order will have a config in reverse order and slightly less complex code.  If you reduce this to graph traversal concepts, there should be no difference in complexity on matching linear patterns in natural vs reverse order, it's just a matter of convenience based on how the graph is implemented.

I think optimizing for a clean, natural order config is the right thing to do here - config readability is more important than code conciseness.

vkuzo replied:
code like this is "in reverse":

```
        if node.op != "call_function" or node.target not in [torch.ops.aten.relu_.default, torch.ops.aten.relu.default]:
            return
        relu_node = node
        conv_node = relu_node.args[0]
        if conv_node.op != "call_function" or conv_node.target != torch.ops.aten.convolution.default:
            return
```

also, if a user is debugging this stuff and they want to print results after node by node (what's the difference between annotating node 0, node 0,1 node 0,1,2, etc), it's a lot easier to debug if things are happening in natural order

vkuzo replied:
> also what's your thoughts on the annotation rule change: we can annotate one node multiple times, and the annotation survived in the end is whichever comes last, and we probably need to run the code to know which annotation survived

IMO there is no relationship from rules like this to whether we traverse in natural or reverse order, in terms of the underlying algorithm.  The correct implementation might be slightly different based on traversal order and how the graph format works, but the concepts should be the same.  I think you may get some of the implementation details "for free" in reverse order, but it should be very doable to implement in natural order.

You could do something like:
1. each node can only be annotated once
2. in each annotating function, the code determines the order of annotation.  For example, if the code matches conv_1-relu_1 first, it won't match just conv_1 later.
3. backends could change (2) if they need to based on what they want to accomplish

jerryzh168 replied:
I see, makes sense, I think I was talking about the case of traversing in natural order but matching in reverse order, while you were talking about doing both traversal and matching in natural order. so these two are related, the rules will be simply if these two are aligned. so right now how do we do matching is undecided, I think Kimish might want to use SubgraphMatcher to do the matching, is it OK if we postpone this until that is decided?

vkuzo replied:
I guess I care about this enough that I commented on a prototype :)

IMO, ranked from best to worst
1. natural order everything
2. reverse order everthing
3. some things are natural order, some things are reverse order

If there is some dependency which has to use reverse order, then I think (2) is better than (3).  However, (1) is the best UEX.  Would love to know how SubgraphMatcher is related, does it have some dependency on order of something?  I haven't used it.

jerryzh168 replied:
actually another change we'd see if we are using SubgraphMatcher is another dimension of traversal, right now we loop through all node and for each node we do some matching, if we use SubgraphMatcher, since it is a whole graph matching, we will not do this node by node traversal, but instead just loop through the patterns we want to match and match the pattern agains the entire graph to do annotation:
```
for pattern in patterns:
    subgraph_matcher.match(patter, model.graph, ...)
    annotate
```
so the logic will be different, this detail is not decided yet right now, so I hope to land something to get this moving and we can fix it after we actually make a decision on this

jerryzh168 replied:
> I guess I care about this enough that I commented on a prototype :)
> 
> IMO, ranked from best to worst
> 
> 1. natural order everything
> 2. reverse order everthing
> 3. some things are natural order, some things are reverse order
> 
> If there is some dependency which has to use reverse order, then I think (2) is better than (3). However, (1) is the best UEX. Would love to know how SubgraphMatcher is related, does it have some dependency on order of something? I haven't used it.

yeah this sounds good to me, if we use node by node matching we can change the code to do natural order everything instead of reverse order everything (current code), if we use SubgraphMatcher, this doesn't apply since it is not node by node matching as I mentioned in my last comment

jerryzh168 commented on torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:
```diff
@@ -235,32 +256,168 @@ def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
+        if _is_annotated([relu_node, conv_node]):
+            return
+
+        conv_node.meta["target_dtype_info"] = {
```
yeah sounds good, we can do that in a followup PR

jerryzh168 committed:
Update on "[quant][pt2e] QNNPackQuantizer support annotation for resnet18"


Summary:
This PR adds annotation support for conv2d relu, linear, maxpool2d, add and add relu so
that we can successfully quantize resnet18 with the prepare_pt2e_quantizer API and get the same result
as fx graph mode quantization

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels.test_resnet18_with_quantizer_api

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo reviewed (approved):
accepting to unblock

jerryzh168 committed:
Update on "[quant][pt2e] QNNPackQuantizer support annotation for resnet18"


Summary:
This PR adds annotation support for conv2d relu, linear, maxpool2d, add and add relu so
that we can successfully quantize resnet18 with the prepare_pt2e_quantizer API and get the same result
as fx graph mode quantization

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels.test_resnet18_with_quantizer_api

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] QNNPackQuantizer support annotation for resnet18"


Summary:
This PR adds annotation support for conv2d relu, linear, maxpool2d, add and add relu so
that we can successfully quantize resnet18 with the prepare_pt2e_quantizer API and get the same result
as fx graph mode quantization

Test Plan:
python test/test_quantization.py TestQuantizePT2EModels.test_resnet18_with_quantizer_api

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot merge
